### PR TITLE
Vortex performance improvement: Enable multiple stream clients per worker

### PR DESCRIPTION
**Pull request by prodriguezdefino**
Changing the StorageWrite stream append client cache to use a list of entries instead of just one entry per stream.

In the majority of the cases the default configuration would be sufficient to address the ingestion volume, but there are cases were the ingestion will need a higher level of parallelism. Setting `--numStorageWriteApiStreams=X` to a value > 1 will help to increase the ingestion parallelism level, users should be mindful of changing this number to a high one since the number of open connections will rapidly scale with the number of workers potentially reaching BigQuery quota/limits (see: https://cloud.google.com/bigquery/quotas#write-api-limits).   

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by prodriguezdefino** [comment]
R: @reuvenlax

**Comment by prodriguezdefino** [comment]
Run Java PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 1]
```diff
@@ -263,9 +295,9 @@ void invalidateWriteStream() {
             // thread has already invalidated
             // and recreated the stream).
             @Nullable
-            StreamAppendClient cachedAppendClient = APPEND_CLIENTS.getIfPresent(streamName);
-            if (cachedAppendClient != null
-                && System.identityHashCode(cachedAppendClient)
+            List<StreamAppendClient> cachedAppendClients = APPEND_CLIENTS.getIfPresent(streamName);
+            if (cachedAppendClients != null
+                && System.identityHashCode(cachedAppendClients.get(clientNumber))
                     == System.identityHashCode(streamAppendClient)) {
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:301]: This isn't quite right - we're now invalidating all of the StreamWriters when any one of them fails. I think instead you want to just null out the one that failed and allow it to be recreated the next get.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 2]
```diff
@@ -357,28 +389,33 @@ void flush(RetryManager<AppendRowsResponse, Context<AppendRowsResponse>> retryMa
...
     private int numPendingRecords = 0;
     private int numPendingRecordBytes = 0;
-    private static final int FLUSH_THRESHOLD_RECORDS = 150000;
     private final int flushThresholdBytes;
+    private final int flushThresholdCount;
     private final StorageApiDynamicDestinations<ElementT, DestinationT> dynamicDestinations;
     private final BigQueryServices bqServices;
     private final boolean useDefaultStream;
+    // default append client count to 1
+    private Integer streamAppendClientCount = 1;
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:398]: why not private int?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 1 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:301]: The invalidation here corresponds to a schema mismatch, shouldn't all the clients be invalidated for a particular stream?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryOptions.java [hunk 3]
```diff
@@ -93,7 +93,10 @@
   void setUseStorageWriteApiAtLeastOnce(Boolean value);
 
   @Description(
-      "If set, then BigQueryIO.Write will default to using this number of Storage Write API streams.")
+      "If set, then BigQueryIO.Write will default to using this number of Storage Write API streams. "
+          + "The number of streams indicated will be allocated at a per worker and destination basis, "
+          + "a high number can cause a large pipeline to go over the BigQuery connection quota quickly. "
+          + "With low-mid volume pipelines using the default configuration should be enough.")
   @Default.Integer(0)
```
[the remark this answers is not in the document]
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryOptions.java:97]: done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 4]
```diff
@@ -213,22 +229,38 @@ String createStreamIfNeeded() {
...
             createStreamIfNeeded();
             synchronized (APPEND_CLIENTS) {
               if (lookupCache) {
                 this.streamAppendClient =
-                    APPEND_CLIENTS.get(
-                        streamName,
-                        () ->
-                            datasetService.getStreamAppendClient(
-                                streamName, descriptorWrapper.descriptor));
+                    APPEND_CLIENTS.get(streamName, () -> generateClients()).get(clientNumber);
```
[the remark this answers is not in the document]
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java]: Instead of making the client creation lazy, I reverted the cache structure back to have a single client per entry. But now, the cache key is the stream name + the assigned client number.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 1 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java]: made the changes to only invalidate the writer in use by the bundle, not all of them.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 2 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryOptions.java [hunk 5]
```diff
@@ -93,12 +93,21 @@
...
   Integer getNumStorageWriteApiStreams();
 
   void setNumStorageWriteApiStreams(Integer value);
 
+  @Description(
+      "The number of stream append clients indicated will be allocated at a per worker and destination "
+          + "basis. A large value can cause a large pipeline to go over the BigQuery connection quota quickly. "
+          + "With low-mid volume pipelines using the default configuration should be enough.")
+  @Default.Integer(1)
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryOptions.java:106]: A bit confusing - need to clarify that this only applies for at-least once writes using the default stream

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteRecordsInconsistent.java [hunk 6]
```diff
@@ -50,14 +49,16 @@ public PCollection<Void> expand(PCollection<KV<DestinationT, StorageApiWritePayl
     BigQueryOptions bigQueryOptions = input.getPipeline().getOptions().as(BigQueryOptions.class);
     // Append records to the Storage API streams.
     input.apply(
-        "Write Records",
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteRecordsInconsistent.java:53]: Changing transform names can affect update compatibility - do you need this?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 7]
```diff
@@ -197,6 +204,10 @@ String getDefaultStreamName() {
         return BigQueryHelpers.stripPartitionDecorator(tableUrn) + "/streams/_default";
       }
 
+      String getStreamAppendClientCacheEntryName() {
+        return getDefaultStreamName() + "-client" + clientNumber;
+      }
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:209]: This is a bit weird, since this code doesn't always use the default stream. Now the cache is probably not needed in the non default stream case (since we'll create a new stream for every bundle), however if we change that we need to rename the cache and also make sure to close the client (since right now we rely on the cache removal listener to close the client)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 8]
```diff
@@ -221,14 +242,11 @@ StreamAppendClient getStreamAppendClient(boolean lookupCache) {
               if (lookupCache) {
                 this.streamAppendClient =
                     APPEND_CLIENTS.get(
-                        streamName,
-                        () ->
-                            datasetService.getStreamAppendClient(
-                                streamName, descriptorWrapper.descriptor));
+                        getStreamAppendClientCacheEntryName(), () -> createStreamAppendClient());
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:245]: getStreamAppendClientCacheEntryName doesn't necessarily return the stream name we used

Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:245]: It will also be broken if you have two sinks in the pipeline, one using the default stream and one not.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteRecordsInconsistent.java (hunk 6 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteRecordsInconsistent.java:53]: No I don't, but this does not follow the same convention other apply labels use (no spaces on names). Should I revert?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 8 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:245]: I can see the problem of mixing the default with dynamically gen streams, I will change that.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 7 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:209]: I will make the changes to consider default streams and per bundle on-demand streams.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 9]
```diff
@@ -409,9 +447,18 @@ private void initializeDatasetService(PipelineOptions pipelineOptions) {
       }
     }
 
+    @Setup
+    public void setup() {
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:451]: I'm not sure what this is adding?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 10]
```diff
@@ -409,9 +447,18 @@ private void initializeDatasetService(PipelineOptions pipelineOptions) {
...
+      if (useDefaultStream) {
+        destinations = Maps.newHashMap();
+      }
+    }
+
     @StartBundle
     public void startBundle() throws IOException {
-      destinations = Maps.newHashMap();
+      if (!useDefaultStream) {
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:459]: Why add this if?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 11]
```diff
@@ -455,21 +506,26 @@ public void process(
     @FinishBundle
     public void finishBundle(FinishBundleContext context) throws Exception {
       flushAll();
-      for (DestinationState state : destinations.values()) {
-        if (!useDefaultStream) {
+      if (!useDefaultStream) {
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:509]: Why this change? Won't this prevent us from unpinning the client?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 12]
```diff
@@ -455,21 +506,26 @@ public void process(
...
-      destinations.clear();
-      destinations = null;
     }
 
     @Teardown
     public void teardown() {
+      if (destinations != null) {
+        for (DestinationState state : destinations.values()) {
+          state.teardown();
+        }
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:527]: We shouldn't be doing this in teardown

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryOptions.java [hunk 13]
```diff
@@ -93,12 +93,22 @@
...
   @Description(
-      "If set, then BigQueryIO.Write will default to using this number of Storage Write API streams.")
+      "If set, then BigQueryIO.Write will default to using this number of Storage Write API streams. ")
   @Default.Integer(0)
   Integer getNumStorageWriteApiStreams();
 
   void setNumStorageWriteApiStreams(Integer value);
 
+  @Description(
+      "When using the \"_default\" table stream, this option sets the number of stream append clients that will be allocated "
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryOptions.java:103]: reference at-least once writes. Users don't know about default streams, as that's an implementation detail

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 9 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:451]: I've seen evidence that closing one stream append client for the `_default` stream causes a cascade close of the other ones, so moving the state for destinations to be reused between bundle executions at least decreased the occurrences of those cascading closes. 

I can revert this if this is not the right idea to try.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 10 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:459]: Only want to reuse the destination state for the default stream clients.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 11 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:509]: it will, but those clients can be reused across bundle executions.

Comment (reuvenlax) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:509]: the clients can be reused regardless - unpinning won't close the client unless close is also called (i.e. if we hit the cache idle timeout)

**Comment by y1chi** [comment]
Run Java PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 14]
```diff
@@ -213,6 +226,14 @@ String createStreamIfNeeded() {
         return this.streamName;
       }
 
+      StreamAppendClient generateClient() {
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:229]: Minor nit - remove try/catch since the calling function already catches the exception

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java [hunk 15]
```diff
@@ -263,11 +281,12 @@ void invalidateWriteStream() {
             // thread has already invalidated
             // and recreated the stream).
             @Nullable
-            StreamAppendClient cachedAppendClient = APPEND_CLIENTS.getIfPresent(streamName);
+            StreamAppendClient cachedAppendClient =
+                APPEND_CLIENTS.getIfPresent(getStreamAppendClientCacheEntryName());
```
Comment (reuvenlax) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java:285]: nit - store cachedEntryName in a local variable instead of recomputing twice here

**Comment by reuvenlax** [comment]
A couple of very minor nits, but otherwise LGTM

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 14 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java]: done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java (hunk 15 above)
Comment (prodriguezdefino) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWriteUnshardedRecords.java]: done

**Comment by prodriguezdefino** [comment]
Run Java PreCommit

**Comment by reuvenlax** [comment]
Run Java PreCommit

**Comment by y1chi** [comment]
Run Java PreCommit
